Cql/Cql.CqlToElm/LibraryProviders/StreamInspector.cs

```diff
@@ -0,0 +1,157 @@
+﻿using Hl7.Cql.Elm;
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+using System.Text;
+using System.Text.Json;
+using System.Text.RegularExpressions;
+using System.Threading.Tasks;
+
+namespace Hl7.Cql.CqlToElm.LibraryProviders
+{
+
+    internal class StreamInspector
+    {
+        public bool TryIdentify(Stream stream, out LibraryInfo? libraryInfo)
+        {
+            var sr = new StreamReader(stream);
+            var first = PeekFirstNonWhitespaceCharacter(sr);
+            switch (first)
+            {
+                case '{':
+                    {
+                        var lib = FromJson(stream);
+                        if (lib is null || lib.identifier is null)
+                        {
+                            libraryInfo = null;
+                            return false;
+                        }
+                        else
+                        {
+                            libraryInfo = new LibraryInfo(lib.identifier, lib, LibraryInfo.ElmJsonMimeType);
+                            return true;
+                        }
+                    }
+                case '<':
+                    {
+                        var lib = FromXml(stream);
+                        if (lib is null || lib.identifier is null)
+                        {
+                            libraryInfo = null;
+                            return false;
+                        }
+                        else
+                        {
+                            libraryInfo = new LibraryInfo(lib.identifier, lib, LibraryInfo.ElmXmlMimeType);
+                            return true;
+                        }
+                    }
+                default:
+                    {
+                        var vi = FromCql(sr);
+                        if (vi is not null)
+                        {
+                            libraryInfo = new LibraryInfo(vi, null, LibraryInfo.CqlMimeType);
+                            return true;
+                        }
+                        else
+                        {
+                            libraryInfo = null;
+                            return false;
+                        }
+                    }
+            }
+        }
+
+        internal Library? FromJson(Stream stream)
+        {
+            try
+            {
+                return Library.LoadFromJson(stream);
+            }
+            catch { return null; }
+        }
+        private Library? FromXml(Stream reader) => null;
+
+        private Regex LibraryWithVersion = new Regex("^library\\s+(?'name'\\S+)(\\s+version\\s+'(?'version'\\d+(\\.\\d+)*)')$", RegexOptions.Compiled);
+        private Regex LibraryWithoutVersion = new Regex("^library\\s+(?'name'\\S+)$", RegexOptions.Compiled);
+        internal VersionedIdentifier? FromCql(StreamReader reader)
+        {
+            // TODO: use ANTLR?
```

Comment:
No, Pidgin (https://github.com/benjamin-hodgson/Pidgin) or Sprache (https://github.com/sprache/Sprache) would be perfect for this task. Let me know, I have worked with these tools before.

Reply:
Then again, Antlr already has the syntactic rules for identifiers and versions too, so that's worth trying out indeed.

Reply:
Turned out to be super easy.